# SVM Loan Approver
There are a number of classification algorithms that can be used to determine loan elgibility. Some algorithms run better than others. Build a loan approver using the SVM algorithm and compare the accuracy and performance of the SVM model with the Logistic Regression model.

In [1]:
#17.5.2 SVM in Practice
from path import Path
import numpy as np
import pandas as pd

In [2]:
df = pd.read_csv(Path('loans.csv'))
df.head()

# # Read in the data
# # Note: The following data has been normalized between 0 and 1
# data = Path('loans.csv')
# df = pd.read_csv(data)
# df.head()

assets  liabilities    income  credit_score  mortgage   status
0  0.210859     0.452865  0.281367      0.628039  0.302682     deny
1  0.395018     0.661153  0.330622      0.638439  0.502831  approve
2  0.291186     0.593432  0.438436      0.434863  0.315574  approve
3  0.458640     0.576156  0.744167      0.291324  0.394891  approve
4  0.463470     0.292414  0.489887      0.811384  0.566605  approve

# Separate the Features (x) from the Target (y)

In [3]:
# Segment the features from the target
y = df["status"]
X = df.drop(columns="status")

# Split our data into training and testing

In [4]:
# Use the train_test_split function to create training and testing subsets
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, 
                                                    y, 
                                                    random_state=1, 
                                                    stratify=y)
X_train.shape

(75, 5)

# Create a SVM Model

In [5]:
# Instantiate a linear SVM model
from sklearn.svm import SVC
model = SVC(kernel='linear')

# Fit (train) or model using the training data

In [6]:
# Fit the data
model.fit(X_train, y_train)

SVC(kernel='linear')

# Score the model using the test data

## Make predictions

In [7]:
# Make predictions using the test data
y_pred = model.predict(X_test)
results = pd.DataFrame({
    "Prediction": y_pred, 
    "Actual": y_test
}).reset_index(drop=True)
results.head()

Prediction   Actual
0    approve     deny
1       deny  approve
2       deny     deny
3    approve     deny
4       deny     deny

In [8]:
from sklearn.metrics import accuracy_score
accuracy_score(y_test, y_pred)

0.6

# Generate Confusion Matrix

In [9]:
    from sklearn.metrics import confusion_matrix
    confusion_matrix(y_test, y_pred)

array([[7, 5],
       [5, 8]], dtype=int64)

# Generate Classification Report

In [10]:
from sklearn.metrics import classification_report
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

     approve       0.58      0.58      0.58        12
        deny       0.62      0.62      0.62        13

    accuracy                           0.60        25
   macro avg       0.60      0.60      0.60        25
weighted avg       0.60      0.60      0.60        25



# 17.5.2 Script

In [11]:
y = df["status"]
X = df.drop(columns="status")

In [12]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X,
   y,  random_state=1, stratify=y)
X_train.shape

(75, 5)

In [13]:
from sklearn.svm import SVC
model = SVC(kernel='linear')

In [14]:
model.fit(X_train, y_train)

SVC(kernel='linear')

In [15]:
SVC(kernel='linear')

SVC(kernel='linear')

In [16]:
y_pred = model.predict(X_test)
results = pd.DataFrame({
   "Prediction": y_pred,
   "Actual": y_test
}).reset_index(drop=True)
results.head()

Prediction   Actual
0    approve     deny
1       deny  approve
2       deny     deny
3    approve     deny
4       deny     deny

In [17]:
from sklearn.metrics import accuracy_score
accuracy_score(y_test, y_pred)

0.6

In [18]:
from sklearn.metrics import confusion_matrix
confusion_matrix(y_test, y_pred)

from sklearn.metrics import classification_report
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

     approve       0.58      0.58      0.58        12
        deny       0.62      0.62      0.62        13

    accuracy                           0.60        25
   macro avg       0.60      0.60      0.60        25
weighted avg       0.60      0.60      0.60        25



# Skill Drill the logistics model was .78 precision and this one is .58 precision, which means the logistics model is mor